### ISSUE TITLE
Object copying using pybind11/functional

### ISSUE THREAD

**Issue by molpopgen** [pybind/pybind11#765]
#### Issue description

In working on wrapping some simulation code with pybind11, I came across a performance roadblock that is due to unnecessary data copies when passing functions in.

The example below mimics the setup of a population, the creation of a function to evaluate a property of individuals, and then another function that records data from the entire population.  The former function (the "fitness function") relies on data being passed in from the population object, and this is where the extra copies are created.

The question is: is there a workaround so that the "fitness function" does not result in extra copies?  In our application, we are wrapping an external library into which these functions are getting passed (that library passes the data into the fitness function).

#### Reproducible example code

I've tested this with the current pybind11 master, GCC5, Ubuntu.  You can also get a full repo from http://github.com/molpopgen/pybind11_copying, in case that is easier.

Please forgive the length--I've tried to minimize this as much as possible while still mocking my real-world scenario.

To head off one question: yes, the fitness function and the "recorder" do the same thing here.  However, in production code, that would not be the case--the fitness function would be called on each individual each generation, and the recorder would typically record something else.  This is just a mock-up to show where copies are occurring and where they are not.

First, we have the C++ part, which sets up the idea of a population and functions to operate on the populations:

```cpp
#include <pybind11/pybind11.h>
#include <pybind11/stl.h>
#include <pybind11/functional.h>
#include <iostream>

using individual = std::pair<std::size_t, std::size_t>;

struct gamete
{
    std::vector<std::size_t> mutations;
    gamete(std::vector<std::size_t> m_) : mutations(std::move(m_)) {}
    gamete(const gamete &g) : mutations(g.mutations)
    {
        std::cout << "gamete got copied\n";
    }
};

struct effect
{
    double s;
    effect(double s_) : s(s_) {}
    effect(const effect &e) : s(e.s) { std::cout << "effect got copied\n"; }
};

struct population
{
    std::vector<individual> ind;
    std::vector<gamete> gametes;
    std::vector<effect> effects;
    population() = default;
    population(const population &p)
        : ind(p.ind), gametes(p.gametes), effects(p.effects)
    {
        std::cout << "population got copied\n";
    }
};

population
make_pop()
{
    population p;
    p.effects.emplace_back(-0.2);
    p.effects.emplace_back(0.4);
    p.gametes.emplace_back(std::vector<std::size_t>{0});
    p.gametes.emplace_back(std::vector<std::size_t>{0,1});
    p.ind.resize(10, { 0, 0 });
    p.ind[3].second=1;
    p.ind[5].first=1;
    return p;
}

using individual_fitness_fxn
    = std::function < double(const individual &, const std::vector<gamete> &,
                             const std::vector<effect> &)>;
using population_recorder_function = std::function<void(const population &)>;

void
process_pop(const population &pop, individual_fitness_fxn fitness,
            population_recorder_function recorder)
{
    for (auto &&i : pop.ind)
        {
            //This will result in gametes + 
            //effects getting copy-constructed
            //and passed to fitness.  This 
            //gets extremely expensive in production
            //code.
            std::cout << "calling fitness fxn:\n";
            fitness(i, pop.gametes, pop.effects);
            std::cout << "done\n";
        }
    //no additional copy is made here.
    std::cout << "calling recorder\n";
    recorder(pop);
    std::cout << "done\n";
}

namespace py = pybind11;

PYBIND11_PLUGIN(copying)
{
    py::module m("copying", "expensive copying");

    py::class_<individual>(m, "Individual")
        .def_readonly("first",&individual::first)
        .def_readonly("second",&individual::second);
    py::class_<gamete>(m, "Gamete")
        .def_readonly("mutations",&gamete::mutations);
    py::class_<effect>(m, "Effect")
        .def_readonly("s",&effect::s);
    py::class_<population>(m, "Population")
        .def_readonly("ind",&population::ind)
        .def_readonly("gametes",&population::gametes)

        .def_readonly("effects",&population::effects);

    m.def("make_pop", &make_pop);
    m.def("process_pop",&process_pop);
    return m.ptr();
}
```

Then, the Python test code:

```py
import copying

p = copying.make_pop()

def fitness(dip,g,e):
    """
    Each call to this function
    results in g,e being copied,
    which gets expensive
    """
    sum=0.
    for gi in g[dip[0]].mutations:
        sum+=e[gi].s
    for gi in g[dip[1]].mutations:
        sum+=e[gi].s
    return sum

class Recorder:
    """
    Calls to this object
    work w/o any extra copies
    """
    def __init__(self):
        self.data=[]
    def __call__(self,pop):
        for i in pop.ind:
            self.data.append(fitness(i,pop.gametes,pop.effects))

record=Recorder()
copying.process_pop(p,fitness,record)
print(record.data)
```

**Comment by wjakob** [comment]
The problem is that you are passing around ``std::vector`` instances, which involves a copy at every C++ <-> Python transition. Pybind11 comes with an alternative approach in ``stl_bind.h`` that avoids this, it's all in the documentation.

**Comment by molpopgen** [comment]
OK, I've implemented that, and it does not solve the problem.  (In the project where we ran into this, we'd already made the containers opaque.)

The new C++ code is here:

https://github.com/molpopgen/pybind11_copying/blob/master/copying.cpp

With the new code, I still get ouput showing that the function arguments are being passed in by value rather than by const reference:

calling fitness fxn:
gamete got copied
gamete got copied
effect got copied
effect got copied
done
calling fitness fxn:
gamete got copied
gamete got copied
effect got copied
effect got copied
done
calling fitness fxn:
gamete got copied
gamete got copied
effect got copied
effect got copied
done
calling recorder
done

**Comment by molpopgen** [comment]
I can simplify it further.  It seems that any time a function works on a class member, that class member is getting copied.

This case is highly simplified compared to what we are dealing with, but shows the same fundamental issue: "subobject" gets copied.

In our real-world case, this even happens if we use pybind11 to define a type returning a bound C++ function object.  I have a feeling that pybind11 is generating functions that are removing the reference from the function signature, and forcing a copy at run-time.

```cpp
#include <pybind11/pybind11.h>
#include <pybind11/stl.h>
#include <pybind11/functional.h>
#include <pybind11/stl_bind.h>
#include <iostream>

struct subobject
{
    int x;
    subobject(int x_) : x(x_) {}
    subobject(const subobject& s) : x(s.x)
    {
        std::cout << "subobject copied\n";
    }
};

struct object
{
    subobject s;
    object(int x) : s(x) {}
    object(const object& o) : s(o.s) { std::cout << "object copied\n"; }
};

using sf = std::function<void(const subobject&)>;
using of = std::function<void(const object&)>;

void
doit(const object& o, sf sf_, of of_)
{
    std::cout << "work on object:\n";
    of_(o);
    std::cout << "done\nwork on subobject:\n";
    sf_(o.s);
    std::cout << "done\n";
}

namespace py = pybind11;

PYBIND11_PLUGIN(copying2)
{
    py::module m("copying2", "expensive copying");

    py::class_<subobject>(m, "Subobject")
        .def(py::init<int>())
        .def_readonly("x", &subobject::x);

    py::class_<object>(m, "Object")
        .def(py::init<int>())
        .def_readonly("s", &object::s);

    m.def("doit", &doit);

    return m.ptr();
}
```

```py
import copying2

o=copying2.Object(4)

def work_on_subobj(s):
    print(s.x)

def work_on_obj(o):
    print(o.s.x)

copying2.doit(o,work_on_subobj,work_on_obj)
```

**Comment by jagerman** [comment]
I'm taking a look (I've reproduced it and written a (failing) test case).

**Comment by molpopgen** [comment]
Thanks. 

A short-term workaround is to declare your functions as taking a const pointer to the nested object.  While that works in the sense that it obviously won't cause a copy, it means that using existing C++ libraries will require some potentially tedious wrapper code, which may cause performance loss if indirections can't get inlined out.  I've not had a chance to test if there is any such loss for my project, but I hope to this week.

**Comment by jagerman** [comment]
I investigated this, and it's not exactly a bug but more an issue with how Python and pybind references interact.

The main issue here is that the C++ Subobject instance that you pass in `sf_(o.s)` isn't a registered instance (that is, no Python instance currently references it).  You can see this (and sort of work around it) by adding `z = os` to your Python code immediately before the `copying2.doit(...)` call: the `subject copied` line disappears from the output.

The main reason is that Python (and thus pybind) already references an instance, which means when we try to cast the `o.s` in the `sf_(o.s)` call, we see the existing reference and create a new Python reference to the existing object.  When we *don't* have an existing reference, pybind is being asked to cast the `const Subject &` into a Python object, and takes the safe path of making a copy of it (we can't reference it because we have no idea how long the reference will remain valid).  So we create a new Python wrapper around a new `Subject` copied from the input value, and pybind manages this new copy, destroying it once it goes out of scope.

Normally, this is where `return_value_policy` comes in: you can cast things with an added `return_value_policy::reference` to tell pybind that it is, in fact, safe to store a reference to the object without making a copy (in which case, it's up to the caller to make sure the reference stays valid).  `functional.h`, however, doesn't currently expose that ability, and it's not clear to me that there is any particularly nice way to add the capability in.

One workaround is to skip `functional.h` and instead take a `py::function` argument, controlling the casting to Python (and return value policies) yourself, i.e. replacing your `doit(...)` with:

```C++
void doit(const object& o, py::function sf, py::function of) {
    std::cout << "work on object:\n";
    of(py::cast(o, py::return_value_policy::reference));
    std::cout << "done\nwork on subobject:\n";
    sf(py::cast(o.s, py::return_value_policy::reference));
    std::cout << "done\n";
}
```

The pointer workaround that you discovered is *not* a valid workaround: the reason it seems to work is that when pybind is asked to cast a pointer to a Python object, it takes ownership of the pointer (by default; though again you can change that with a return value policy), which means it destroys it when the Python object gets garbage collected.  But at that point you're in Undefined Behaviour territory: it's going to try to destruct and `delete` that pointer eventually, but you definitely don't want either of those to happen.  So basically, even with a pointer, you *still* have to use `return_value_policy::reference` so that pybind knows that it can safely reference it but shouldn't manage it.

As for why the `.def_readonly` doesn't change this, it's handled dynamically: we don't actually create a python instance for the property until you actually use it.

**Comment by jagerman** [comment]
@wjakob, @dean0x7d - an extra thought for discussion: it should, in theory, be possible to make automatic casting of members work; the logic is essentially that if we try to cast something with a pointer value in `[R, R+S)` (where R is some registered instance pointer and S is that instance's size) we ought to be able to treat `return_value_policy::automatic` as `return_value_policy::reference_internal`: we know we have an internal reference.

It might be expensive to actually implement that logic, however.

**Comment by molpopgen** [comment]
Thanks--I had a feeling that this wouldn't be straightforward.

I'll mull this over this week. The heads-up on the pointer issue is appreciated.  It also turned out to be 10x slower than it "should" be, in addition to having the UB issue.

I'm motivated to find a solution.  This is the only road block to being able to replace 1000s of lines of Cython with a few 100 lines of C++11.  (Thanks for enabling that, btw--pybind11 has come a long way!)

I'll post what I come up with here, once I find something that also works for the main project.

**Comment by molpopgen** [comment]
I made some progress via the py::function example.  Given that I'm working with some existing code, I wanted to make sure that it works with a C++ callable:

```cpp
struct process_subobject
{
    inline void
    operator()(const subobject& o) const
    {
        std::cout << "cpp callable doing the job!\n";
        std::cout << o.x << '\n';
    }
};

//skipping steps....
    
// This appears automagically castable
// to py::function   
py::class_<process_subobject>(m, "ProcessSubobject")
.def(py::init<>())
.def("__call__", [](const process_subobject& po,const subobject& s) -> void { return po(s); });

void
doit(const object& o, py::function sf_, py::function of_)
{
    std::cout << "work on object:\n";
    of_(py::cast(o, py::return_value_policy::reference));
    std::cout << "done\nwork on subobject:\n";
    sf_(py::cast(o.s, py::return_value_policy::reference));
    std::cout << "done\n";
}
```

That's great.  Your suggestion works well there.

If I make the example more realistic, meaning that we're sending these functions in to an existing C++ template library, I arrive at my current challenge.  Our "subobjects" get copied and passed into our functions. (What is happening individuals getting copied in the simulation to be used as "parents", and the existing individuals get over-written as offspring.)  

Here's what I'm talking about (simplified to avoid the copying in this specific case):

```cpp
#include <pybind11/pybind11.h>
#include <pybind11/stl.h>
#include <pybind11/functional.h>
#include <pybind11/stl_bind.h>
#include <iostream>

namespace py = pybind11;

struct subobject
{
    int x;
    subobject(int x_) : x(x_) {}
    subobject(const subobject& s) : x(s.x)
    {
        std::cout << "subobject copied\n";
    }
};

struct object
{
    subobject s;
    object(int x) : s(x) {}
    object(const object& o) : s(o.s) { std::cout << "object copied\n"; }
};

using sf = std::function<void(const subobject&)>;
using of = std::function<void(const object&)>;

struct process_subobject
{
    inline void
    operator()(const subobject& o) const
    {
        std::cout << "cpp callable doing the job!\n";
        std::cout << o.x << '\n';
    }
};

template <typename sf_>
void
external_lib_template_fxn(object& o, const sf_ sf)
{
    /* The "real" project
     * would work on a copy
     * of o.s, but this
     * is sufficient to
     * show that there's still a copy
     */
    std::cout << "about to call our policy\nin external library...\n";
    sf(o.s);
    std::cout << "done with policy call\n";
}

void
doit(object& o, py::function sf_, of of_)
{
    std::cout << "pass off to another lib\nthat works on the subobject\n";
    external_lib_template_fxn(o, sf_);
    std::cout << "work on object in our Python package:\n";
    of_(o);
}

PYBIND11_PLUGIN(copying4)
{
    py::module m("copying4", "expensive copying");

    py::class_<subobject>(m, "Subobject")
        .def(py::init<int>())
        .def_readonly("x", &subobject::x);

    py::class_<object>(m, "Object")
        .def(py::init<int>())
        .def_readonly("s", &object::s);

    // This appears automagically castable
    // to py::function
    py::class_<process_subobject>(m, "ProcessSubobject")
        .def(py::init<>())
        .def("__call__", [](const process_subobject& po,
                            const subobject& s) -> void { return po(s); });

    m.def("doit", &doit);

    return m.ptr();
}
```

And the script:

```py
import copying4

o=copying4.Object(4)

def work_on_subobj(s):
    print(s.x)

def work_on_obj(o):
    print(o.s.x)

po = copying4.ProcessSubobject()

copying4.doit(o,po,work_on_obj)
```

I will note that I can get rid of the subobject copy in the "library" code by modifying the python script as follows:

```py
s=o.s
copying4.doit(o,po,work_on_obj)
```

So, that's great, but that does not model the behavior of the actual library (which copies input data).

My current thought is to force users to write these custom functions using C++11/pybind11, and have those objects follow an API requiring that they can return a properly-bound object to the C++ code, such that pybind11 never has to see any of it.  This may not be so bad for right now, as all other extensions could be written in pure Python and not suffer from these copying issues (insofar as I've tested things...).  Most other extensions would model "work_on_object".

**Comment by dean0x7d** [comment]
> @jagerman said:
> The pointer workaround that you discovered is not a valid workaround

It is valid here. The default `return_value_policy` for the `def` family of wrappers is `automatic` (copy/move values and references; take ownership of pointers) but almost everything else like `py::cast`, `object::operator()`, `py::make_tuple`, etc. take arguments as `automatic_reference` (copy/move values and references; don't own any pointers). This seemed a bit weird to me at first, but it does make for a nice terse notation:

```c++
void f(py::function foo) {
    auto x = Widget{};
    foo(x); // copy object
    foo(&x); // no copy, just reference -- shorter form of the line below
    foo(py::cast(x, py::return_value_policy::reference)); // also reference, but verbose
}
```

As mentioned, this kind of selection isn't possible with `std::function<Return(Args...)>` because both the return value policy and the argument type (value/reference vs. pointer) are fixed so it ends up copying everything except pointers (that's a sensible default). A possible solution is to let users define template specializations which would just say that, e.g. `std::function<void(const population &)>` should use `rvp::reference` instead of the default `rvp::automatic_reference`. That would solve the copying issue here without any need to change the original C++ code. However:

> @molpopgen said:
> The heads-up on the pointer issue is appreciated. It also turned out to be 10x slower than it "should" be, in addition to having the UB issue.

The pointer solution didn't have any UB issues as mentioned above. But if you still found that to be slow then there is probably some other issue besides the copying. You mentioned it's not a fast as it "should" be. What's the comparison point there? Perhaps it's just the C++ <-> Python call overhead.

**Comment by molpopgen** [comment]
> The pointer solution didn't have any UB issues as mentioned above. But if you still found that to be slow then there is probably some other issue besides the copying. You mentioned it's not a fast as it "should" be. What's the comparison point there? Perhaps it's just the C++ <-> Python call overhead.

I've created a py::class_ that wraps an existing function object, so I can compare the speed of using that function object alone to the speed of a wrapper that takes objects and sends pointers to them.

In other words, using the library without pybind11 is pretty normal:

```cpp
library_fxn(std::bind(fxn_obj,std::placeholders::_1_thru_n)
```

But when using pointers, I created a wrapper object (not wrapped by pybind11, but intended as glue to the c++ library I'm using):

```cpp
/*
This is defined in a header, and all the types below are wrapped in py::class_, and
the latter 2 are opaque containers.
*/
using singlepop_fitness = std::function<double(const fwdpy11::diploid_t *,
                                                   const fwdpy11::gcont_t *,
                                                   const fwdpy11::mcont_t *)>;

//This is not run thru py::class_, 
//and is used to translate a callable
//taking pointers to one taking references
struct fitness_model_sink
{ 
       const fwdpy11::singlepop_fitness f;
        explicit fitness_model_sink(fwdpy11::singlepop_fitness f_) : f(std::move(f_))
        {
        }
        inline double operator()(const fwdpy11::dipcont_t & diploid,
                        const fwdpy11::gcont_t & gametes, 
                        const fwdpy11::mcont_t & mutations) const
        {
                return f(&diploid,&gametes,&mutations);
        }
};
```

The above struct would be used like:

```cpp
//This is exposed to Python using pybind11
void function( const fwdpy11::singlepop_fitness f, other_params...)
{
    fitness_model_sink sink(f);

    //... refers to there being other 
    //params involved
    other_library_function(...,f...);
}
```

My comment re: speed is that I can mock a case where I know what "f" is doing (referring to an existing functor), and simply drop in a std::bind(functor_type(),etc.), and I find and 8-10x speedup.

But, point taken that I've not shown there isn't another unexpected copy being generated somewhere.

I'm not sure that there should be a lot of Py -> Cpp conversion going on here. (That may be wishful thinking, though.)  This callable should get applied to every individual over a long number of generations, meaning easily millions of calls.  But this is all happening on the C++ side, in the library evolving the population.

**Comment by molpopgen** [comment]
@dean0x7d @jagerman 

First, thanks for your input.  I have a much better understanding of what is doing on now.

The TL;DR is that what I'm trying to do is unsupported, but it can be forced to work (see below).

A fundamental difficulty is that the one of the arguments (and individual) coming into the callback are constantly changing, meaning the whole idea of a Python process keeping references to them is simply absurd.

That said, I came up with a means of wrapping the type of function object that my library needs while also avoiding copies of objects, which I demonstrated to myself using simple cout lines from the const object constructors for the relevant types.  I am sure that this is unsafe, unsupported, fragile, and likely prone to break if any change to pybind11 ever occurs :).  I'm just going to copy/paste what worked in my app, rather than a simplified example:

```cpp
struct sink
{
    py::function f;
    sink(py::function f_) : f(std::move(f_)) {}
    //This fxn sig is what the external C++11 lib requires.  It is templated,
    //but the type constraints are well-defined:
    inline double
    operator()(const fwdpy11::diploid_t& dip, const fwdpy11::gcont_t& gametes,
               const fwdpy11::mcont_t& mutations) const
        /* The call operator returns a double never seen by Python.
         * It is recorded internally in the c++ library we're using.
         * The arguments are passed in by that lib, too.
         */
    {
                //steal pointers to objects
                //and avoid making copies
        auto d = py::cast(&dip, py::return_value_policy::automatic_reference);
        auto g
            = py::cast(&gametes, py::return_value_policy::automatic_reference);
        auto m = py::cast(&mutations,
                          py::return_value_policy::automatic_reference);
                //My compiler says a py::function cannot return a double,
                //so we get the rv...
        auto rv
            = f(*d.cast<fwdpy11::diploid_t*>(), *g.cast<fwdpy11::gcont_t*>(),
                *m.cast<fwdpy11::mcont_t*>());
                //...and cast it
        return rv.cast<double>();
    }
};
```

So, it works.  Yay!  But, aside from likely having side-effects that I'm unaware of, it is no faster than versions based on previous posts where I fool pybind11 into holding a reference to "gametes" and "mutations" while allowing the "diploid" to be copied at each call.  In other words, it is at least 10x slower than the native C++ code that it is wrapping.

I think I'm going to go to Plan B here, which is to define a base class with a pure virtual function returning a bound c++ callable.  That function returning the callable will __not__ be exposed to pybind11.  Rather, it will only be called from the pure C++ side, and thus these functions will have to be C++ only.  User's "custom" callables will inherit from this base class, and a const reference to the base will be passed in to the C++ functions.

I'm going to be happy with this: it is one thing that'll require C++ rather than tons of things requiring Cython.  It'll also allow "stateful" callables because std::bind will be allowed, which Cython cannot handle.

Again--thanks for the insight.  I'm going to start on Plan B this week, giving you until Monday or so to talk me out of it :).

**Comment by jagerman** [comment]
Your workaround is essentially doing exactly what I was doing in Python by holding the instances alive in `d`, `g`, and `m` (that you're doing this from C++ instead of Python is irrelevant: they are still wrapped, and still valid references).  Since they're living references, pybind will pick them up in the later call and avoid a copy.

In other words, what you have there won't stop working in future pybind11 releases.

**Comment by molpopgen** [comment]
@jagerman 
> Your workaround is essentially doing exactly what I was doing in Python by holding the instances alive in d, g, and m (that you're doing this from C++ instead of Python is irrelevant: they are still wrapped, and still valid references). Since they're living references, pybind will pick them up in the later call and avoid a copy.
> 
> In other words, what you have there won't stop working in future pybind11 releases.

Ok, that's good to know.  I can imagine there being other cases where this trick will be useful and it isn't as performance-critical a section of code.

**Comment by molpopgen** [comment]
Two last things:

1. Plan B worked great.
2. The one thing about my C++ trick is that it relies on `py::return_value_policy::automatic_reference`, which the docs say you "probably won't need".  While that is certainly still  true, this does appear to be one use case that could be mentioned.

I'll leave it up to you to close the issue, or if you think there should be some update to the docs re: how to deal nicely with object members of classes or whatnot.

Thanks again for the help.

**Comment by jagerman** [comment]
You "probably won't need" in this case in the sense that it is always resolves to `return_value_policy::reference` when given a pointer, so you could just specify that directly instead.
